cohorts/cohort.py

```diff
@@ -470,23 +489,117 @@ def load_variants(self, patients=None, filter_fn=None):
+            kw_hash = ['default']
+        else:
+            [kw_hash.append('{}-{}'.format(key, h)) for (key, h) in sorted(kw_dict.items())]
+        # closure vars
+        closure = 'null'
+        nonlocals = inspect.getclosurevars(filter_fn).nonlocals
+        for (key, val) in nonlocals.items():
+            if inspect.isfunction(val):
+                closure = '{}-{}'.format(self._hash_filter_fn(val).__name__, self._hash_filter_fn(val))
+        # construct final hashed_fn
+        hashed_fn = '.'.join(['-'.join([filter_fn.__name__,
+                                        int(hashlib.sha1(pickled_fn_source).hexdigest(), 16) % (10 ** 11)]),
+                              '.'.join(kw_hash),
+                              closure]
+                            )
+        return hashed_fn
+            
+        
+    
+    def _load_single_patient_variants(self, patient, filter_fn, use_cache=True, **kwargs):
+        if filter_fn is None:
+            filter_fn_name = 'None'
+        else:
+            filter_fn_name = filter_fn.__name__
+        logger.debug('loading variants for patient {} with filter_fn {}'.format(patient.id, filter_fn_name))
+        use_filtered_cache = use_cache
+        if sys.version_info < (3, 3):
+            logger.debug('... disabling filtered cache due to python version')
+            use_filtered_cache = False
+            
+        ## get cache name, if possible
+        if use_filtered_cache:
+            logger.debug('... identifying filtered-cache file name')
+            try:
+                ## try to load filtered variants from cache
+                filtered_cache_file_name = "%s-variants.%s.pkl" % (self.merge_type,
+                                                                          self._hash_filter_fn(filter_fn, **kwargs))
+            except:
+                use_filtered_cache = False
+        
+        ## read from cache, if possible
+        if use_filtered_cache:
+            logger.debug('... trying to load filtered variants from cache: {}'.format(filtered_cache_file_name))
+            try:
+                cached = self.load_from_cache(self.cache_names["variant"], patient.id, filtered_cache_file_name)
+                if cached is not None:
+                    return cached
+            except:
+                logger.warn('Error loading variants from cache for patient: {}'.format(patient.id))
+                pass
+        
+        ## get merged variants
+        logger.debug('... getting merged variants for: {}'.format(patient.id))
+        merged_variants = self._load_single_patient_merged_variants(patient, use_cache=use_cache)
 
-    def _load_single_patient_variants(self, patient, filter_fn):
+        # Note that this is the number of variant collections and not the number of
+        # variants. 0 variants will lead to 0 neoantigens, for example, but 0 variant
+        # collections will lead to NaN variants and neoantigens.
+        if merged_variants is None:
```

Comment:
I re-edited it to clarify that we want to preserve the value of `None` as being qualitatively different from `0`.